gslib/utils/shim_util.py

```diff
@@ -0,0 +1,219 @@
+        field must be a dict of sub-command-name:GcloudStorageMap pairs.
+      flag_map (dict): A dict of str to GcloudStorageFlag. Mapping of gsutil
+        flags to their equivalent gcloud storage flag names.
+      supports_output_translation (bool): Indicates if the corresponding
+        gcloud storage command supports the printing gsutil formatted output.
+    """
+    self.gcloud_command = gcloud_command
+    self.flag_map = flag_map
+    self.supports_output_translation = supports_output_translation
+
+
+def _get_gcloud_binary_path():
+  cloudsdk_root = os.environ.get('CLOUDSDK_ROOT_DIR')
+  if cloudsdk_root is None:
+    raise exception.GcloudStorageTranslationError(
+        'Gcloud binary path cannot be found.')
+  return os.path.join(cloudsdk_root, 'bin', 'gcloud')
+
+
+class GcloudStorageCommandMixin(object):
+  """Provides gcloud storage translation functionality.
+  
+  The command.Command class must inherit this class in order to support
+  converting the gsutil command to it's gcloud storage equivalent.
+  """
+  # Mapping for translating gsutil command to gcloud storage.
+  gcloud_storage_map = None
+
+  def __init__(self):
+    self._translated_gcloud_storage_command = None
+    self._translated_env_variables = None
+
+  def _get_gcloud_storage_args(self, sub_opts, gsutil_args, gcloud_storage_map):
+    if gcloud_storage_map is None:
+      raise exception.GcloudStorageTranslationError(
+          'Command "{}" cannot be translated to gcloud storage because the'
+          ' translation mapping is missing.'.format(self.command_name))
+    args = []
+    if isinstance(gcloud_storage_map.gcloud_command, str):
+      args = gcloud_storage_map.gcloud_command.split()
+    elif isinstance(gcloud_storage_map.gcloud_command, dict):
+      # If a command has sub-commands, e.g gsutil pap set, gsutil pap get.
+      # All the flags mapping must be present in the subcommand's map
+      # because gsutil does not have command specific flags
+      # if sub-commands are present.
+      if gcloud_storage_map.flag_map:
+        raise ValueError(
+            'Flags mapping found at command level for the command: {}.'.format(
+                self.command_name))
+      sub_command = gsutil_args[0]
+      sub_opts, parsed_args = self.ParseSubOpts(
+          args=gsutil_args[1:], should_update_sub_opts_and_args=False)
+      return self._get_gcloud_storage_args(
+          sub_opts, parsed_args,
+          gcloud_storage_map.gcloud_command.get(sub_command))
+    else:
+      raise ValueError('Incorrect mapping found for "{}" command'.format(
+          self.command_name))
+
+    if sub_opts:
+      for option, value in sub_opts:
+        if option not in gcloud_storage_map.flag_map:
+          raise exception.GcloudStorageTranslationError(
+              'Command option "{}" cannot be translated to'
+              ' gcloud storage'.format(option))
+        args.append(gcloud_storage_map.flag_map[option].gcloud_flag)
+        if value != '':
```

Comment:
Why not `if value`?

Reply:
The value is going to be empty string if it is not defined as parsed by the ParseSubOpts. So this check makes it more specific.

Reply:
If you're checking for a specific type of falsey value, I would comment why